### ISSUE TITLE
There are no 'Private' subnet groups in this VPC. Available types: Public

### ISSUE THREAD

**Issue by picosam** [aws/aws-cdk#7062]
I'm trying to create a new RDS instance in an isolated subnet of a VPC that I'm creating.

### Reproduction Steps

```
const rdsVpc = new ec2.Vpc(this, 'MyVpc', {
	natGateways: 0,
	subnetConfiguration: [
		{
			cidrMask: 28,
			name: 'rds',
			subnetType: ec2.SubnetType.ISOLATED
		}
	]
});

const dbInstance = new rds.DatabaseInstance(this, 'MyRdsInstance', {
	engine: rds.DatabaseInstanceEngine.POSTGRES,
	instanceClass: new ec2.InstanceType('t3.micro'),
	masterUsername: 'admin',
	databaseName: 'mydb',
	masterUserPassword: this.dbSecretManager.secretValue,
	vpcPlacement: {
		subnetType: SubnetType.ISOLATED
	},
	vpc: rdsVpc
});
```

### Error Log

There are no 'Private' subnet groups in this VPC. Available types: Public
### Environment

  - **CLI Version      : 1.31.0**
  - **Framework Version: **
  - **OS               : MacOS Catalina**
  - **Language         : Typescript**

--- 

This is :bug: Bug Report

**Comment by DerkSchooltink** [comment]
I cannot reproduce your problem with the provided snippet. @picosam 

Deploying a stack with only the VPC with isolated subnets and an RDS instance that uses the isolated subnet seems to work fine:

```
import {App, Stack, StackProps} from "@aws-cdk/core";
import {InstanceType, SubnetType, Vpc} from "@aws-cdk/aws-ec2";
import {DatabaseInstance, DatabaseInstanceEngine} from "@aws-cdk/aws-rds";

export class Test extends Stack {
    constructor(scope: App, id: string, props?: StackProps) {
        super(scope, id, props);

        const rdsVpc = new Vpc(this, 'MyVpc', {
            natGateways: 0,
            subnetConfiguration: [{
                cidrMask: 28,
                name: 'rds',
                subnetType: SubnetType.ISOLATED
            }]
        });

        const dbInstance = new DatabaseInstance(this, 'MyRdsInstance', {
            engine: DatabaseInstanceEngine.POSTGRES,
            instanceClass: new InstanceType('t3.micro'),
            masterUsername: 'jeff',
            databaseName: 'mydb',
            vpcPlacement: {
                subnetType: SubnetType.ISOLATED
            },
            vpc: rdsVpc
        });
    }
}
```

![image](https://user-images.githubusercontent.com/22047701/78004622-a6aeea80-733a-11ea-809f-e6db5994ee42.png)

I am on the exact same setup as well, CDK 1.31.0, OSX Catalina and using TypeScript. Perhaps the error is unrelated to the VPC and RDS instance? Can you do some further investigation?

**Comment by picosam** [comment]
Now that's interesting... ok, I'll try investigating further then.

**Comment by rix0rrr** [comment]
I have a hard time seeing how the code you've posted would produce the error message you've posted.

Will keep this open for a little bit longer, but please let me know how you're getting on.

**Comment by picosam** [comment]
Ok, so this was not related to the creation of the RDS instance but the network balancer instead. Trying to call `ApplicationLoadBalancedFargateService` with the above configuration causes that error.

**Comment by rix0rrr** [comment]
Could you post a small repro please? Still confused how a VPC configured to only have Isolated subnets could throw "Available type: Public".

**Comment by gdamjan** [comment]
> Ok, so this was not related to the creation of the RDS instance but the network balancer instead. Trying to call `ApplicationLoadBalancedFargateService` with the above configuration causes that error.

You can recreate the problem even if you just call `new FargateService(…)` (with a cluster created with the VPC) but it'll work if you add `assignPublicIp: true,` to the `FargateServiceProps`.

The error is quite misleading though:
```
node_modules/@aws-cdk/aws-ec2/lib/vpc.ts:427
      throw new Error(`There are no '${subnetType}' subnet groups in this VPC. Available types: ${availableTypes}`);
            ^
Error: There are no 'Private' subnet groups in this VPC. Available types: Public
```

**Comment by ivawzh** [comment]
Is this issue actually fixed? I am still seeing this error from cdk `1.35.0 (build e0810c8)`

code

```ts
    this.vpc = new Vpc(this, 'Vpc', {
      cidr: '10.180.0.0/16',
      subnetConfiguration: [
        { cidrMask: 22, name: 'Public', subnetType: SubnetType.PUBLIC },
        { cidrMask: 22, name: 'Isolated', subnetType: SubnetType.ISOLATED, reserved: true },
      ]
    })
    this.vpc.addS3Endpoint('S3Endpoint')
    this.vpc.addDynamoDbEndpoint('DynamoDbEndpoint')
```

**Comment by rfpedrosa** [comment]
@ivawzh remove `reserved: true` and it will work ;)

**Comment by evan-beatdapp** [comment]
If anyone stumbles on this and nothing above fixed it - try deleting `cdk.context.json`.

**Comment by bcteam2020** [comment]
vpcSubnets:{subnetType: ec2.SubnetType.ISOLATED}, error fixed (rds along with alb working)

**Comment by mikelhamer** [comment]
Ugh...facing a similar issue of "There are no 'Public' subnet groups in this VPC. Available types: Private". The suggested solutions here are bringing me no joy.

In my case I'm declaring a network balanced service on an existing VPC that has no Public subnets. I don't understand why something is trying to use 'Public'

``` typescript
    const service = new ecspatterns.NetworkLoadBalancedFargateService(this, 'Service', {
      cluster,
      taskSubnets: {subnetType: SubnetType.PRIVATE_WITH_NAT},
      taskDefinition
    });
```

**Comment by mikelhamer** [comment]
Aha!!! Turns out I had to explicitely declare the Network Load Balancer like so. This feels like a bug with NetworkLoadBalancedFargateService. Isn't the whole point of the pattern to avoid having to declare the NLB?

``` typescript
    // Share the load...
    const loadBalancer = new NetworkLoadBalancer(this, 'NetworkLoadBalancer', {
      vpc,
      vpcSubnets: {subnetType: SubnetType.PRIVATE_WITH_NAT}
    });

    const service = new ecspatterns.NetworkLoadBalancedFargateService(this, 'Service', {
      cluster,
      taskDefinition,
      loadBalancer
    });
```

**Comment by github4es** [comment]
I am getting this error `jsii.errors.JSIIError: There are no 'Public' subnet groups in this VPC. Available types: `

```
vpc = Vpc.from_vpc_attributes(self, 'vpc', vpc_id=config.vpc_id,
                                      availability_zones=config.vpc_azs)
```
Any ideas? cdk v1.121.0

**Comment by hoang-innomizetech** [comment]
I am also facing this issue, I am trying to find whether we have a way to receive subnets from VPC? I have two stacks, one is to provision VPC, and another one is to share subnets using RAM with vpc as input. I used the code below

```
props.vpc.selectSubnets({
        subnetGroupName: 'Public',
      }).subnetIds
```

but I am getting below error
```
There are no subnet groups with name 'Public' in this VPC. Available names: 

```

Access to some publicSubnets or privateSubnets also return empty arrays

**Comment by shiromoto-jun** [comment]
I was looking for a solution for "There are no'Public' subnet groups in this VPC. Available types:" and got here.

If you create the VPC and subnet in a separate stack, vpc.publicSubnets and vpc.privateSubnets will return an empty array.
Do you need a special way to register a subnet in your VPC?

**Comment by robertd** [comment]
I'm trying to migrate some old repos (RDS oriented) from cdk 1.x to 2.x and I've been greeted with this error: `There are no 'Private' subnet groups in this VPC. Available types: Isolated,Deprecated_Isolated`. 

```ts
const vpc = ec2.Vpc.fromLookup(this, "vpc", { vpcId: props.config.vpcId });
```

Reverting back to cdk 1.x works just fine. Weird.

**Comment by dillonharless18** [comment]
> I'm trying to migrate some old repos (RDS oriented) from cdk 1.x to 2.x and I've been greeted with this error: `There are no 'Private' subnet groups in this VPC. Available types: Isolated,Deprecated_Isolated`.
> 
> ```ts
> const vpc = ec2.Vpc.fromLookup(this, "vpc", { vpcId: props.config.vpcId });
> ```
> 
> Reverting back to cdk 1.x works just fine. Weird.

This is my issue as well. Can't revert back to CDK V1 given internal requirements...

**Comment by AjkayAlan** [comment]
Somewhere between CDK v1 and v2 CDK changed the way it classified subnets, so things may be a bit wonky. One thing you may want to try locally is:

1. Install CDKv2
2. Clear your context if you have it checked in to your source control/have it in your repo
3. Regenerating it by logging into your AWS CLI with your target AWS account
4. Rerun a synth
5. See how your subnets are classified

```shell
cdk context --clear
# login to your account with the AWS CLI. Below assumes your using the default profile
cdk synth
cat cdk.context.json
```

Within that context it should provide more information about how that subnet is classified by CDK now. Depending on that, you may need to adjust your code when upgrading to CDK v2 to use the new subnet classification

Note - I don't generally recommend clearing context, but this may help you understand things.

**Comment by dillonharless18** [comment]
@AjkayAlan Thank you. In my V1 code I had the subnet on my rds cluster set to PRIVATE (which is now deprecated) and my options in the V2 code were PRIVATE_ISOLATED, PRIVATE_WITH_NAT, and PUBLIC. Setting it to PRIVATE_ISOLATED and running `cdk diff` gave me this error. 

I deleted the context file and tried to run `cdk diff` again but the same thing happened.

I was able to get around it by replacing `subnetType: ec2.SubnetType.PRIVATE_ISOLATED` with `subnets: vpc.privatesubnets`

**Comment by EhsanHerai** [comment]
> @AjkayAlan Thank you. In my V1 code I had the subnet on my rds cluster set to PRIVATE (which is now deprecated) and my options in the V2 code were PRIVATE_ISOLATED, PRIVATE_WITH_NAT, and PUBLIC. Setting it to PRIVATE_ISOLATED and running `cdk diff` gave me this error.
> 
> I deleted the context file and tried to run `cdk diff` again but the same thing happened.
> 
> I was able to get around it by replacing `subnetType: ec2.SubnetType.PRIVATE_ISOLATED` with `subnets: vpc.privatesubnets`

what is `vpc` in here ?

**Comment by AntoniusGolly** [comment]
@EhsanHerai 

I assume, a typo. It should be 

```
vpcSubnets: {
    subnets: vpc.privateSubnets,
  },
```
so, capital "S".

`vpc` being here your VPC. You can reference it with

```
// VPC
const vpc = ec2.Vpc.fromLookup(stack, 'vpc-id', {
  vpcId: buildConfig.Parameters.VpcId,
})
```
I get mine from build config. But, whatever works for you.

Solution worked for me.  @dillonharless18 thanks alot! 🙏

**Comment by 0cv** [comment]
Some mentioned that the `ApplicationLoadBalancer` and the `FargateService` need to have a defined `vpcSubnets`. That was true in my case, and additionally, I also needed to set a `subnets` for the `VpcLink`. 

To troubleshoot the error, I've also found that if you look at the stack trace of that error, it will show the service which is the culprit. It's always ending with `....vpc.js`, but then going up the stack trace, the next one was `vpc-link.js`
